src/com/amazon/ion/IonCursor.java

```diff
@@ -0,0 +1,103 @@
+package com.amazon.ion;
+
+import java.io.Closeable;
+
+/**
+ * A cursor over a stream of Ion data. NOTE: this interface is not considered part of the public API.
+ */
+public interface IonCursor extends Closeable {
+
+    /**
+     * Conveys the type of event that occurred as a result of operating on the cursor.
+     */
+    enum Event {
+
+        /**
+         * There is not enough data in the stream to complete the requested operation. The operation should be retried
+         * when more data is available.
+         */
+        NEEDS_DATA,
+
+        /**
+         * The cursor has completed an operation (e.g. `stepIntoContainer()`) and requires another instruction in order
+         * to position itself on the next value.
+         */
+        NEEDS_INSTRUCTION,
```

Comment:
This seems to be used to mean something else in two places: that the request could not be satisfied due to a buffer configuration limitation. Can you explain that more? What do you expect the user to provide as an instruction in that case?

I wonder if, by pulling the thread on the above questions we _might_ see that there is a different factoring (ala what Matt is poking at) where the range of returns for different requests is more distinct. Specifically that _all_ operations may return `NEEDS_DATA` but are otherwise actually distinct.